windows/src/engine/kmcomapi/com/system/keymancontrol.pas

```diff
@@ -364,14 +379,231 @@ function TKeymanControl.CurrentUILanguage: string;
 end;
 
 destructor TKeymanControl.Destroy;
+begin
+{$IFNDEF WIN64}
+  Keyman32Bit_Destroy;
+{$ENDIF}
+  inherited Destroy;
+end;
+
+procedure TKeymanControl.OpenConfiguration;
+begin
+  RunKeymanConfiguration('');
+end;
+
+procedure TKeymanControl.OpenDiagnostics;
+var
+  msg, path: string;
+begin
+  path := TKeymanPaths.KeymanEngineInstallPath(TKeymanPaths.S_TSysInfoExe);
+  ExecuteProgram('"'+path+'"', ExtractFileDir(path), msg);
+end;
+
+procedure TKeymanControl.OpenHelp(const Topic: WideString);
+begin
+  RunKeymanConfiguration('-h "'+Topic+'"');
+end;
+
+procedure TKeymanControl.OpenTextEditor;
+begin
+  RunKeymanConfiguration('-t');
+end;
+
+procedure TKeymanControl.OpenUpdateCheck;
+begin
+  RunKeymanConfiguration('-ouc');
+end;
+
+procedure TKeymanControl.StartKeyman;
+begin
+  Context.Controller.StartKeyman;
+end;
+
+function TKeymanControl.RunKeymanConfiguration(const filename: string): boolean;
+var
+  err: string;
+begin
+  Result := ExecuteProgram('"'+TKeymanPaths.KeymanDesktopInstallPath(TKeymanPaths.S_KMShell)+'" '+filename,
+    TKeymanPaths.KeymanDesktopInstallDir, err);
+end;
+
+procedure TKeymanControl.Refresh;
+begin
+  if Assigned(FKeymanCustomisation) then
+    FKeymanCustomisation.Refresh;
+end;
+
+procedure TKeymanControl.RefreshWndProc(var Message: TMessage);
+begin
+  with Message do
+    Result := DefWindowProc(RefreshHandle, Msg, WParam, LParam);
+end;
+
+procedure TKeymanControl.DisableUserInterface;
+const
+  KM_DISABLEUI = 1;
+begin
+  if wm_keyman = 0 then
+    wm_keyman := RegisterWindowMessage('wm_keyman');
+  PostMessage(HWND_BROADCAST, wm_keyman, KM_DISABLEUI, 0);
+end;
+
+procedure TKeymanControl.EnableUserInterface;
+const
+  KM_ENABLEUI = 2;
+begin
+  if wm_keyman = 0 then
+    wm_keyman := RegisterWindowMessage('wm_keyman');
+  PostMessage(HWND_BROADCAST, wm_keyman, KM_ENABLEUI, 0);
+end;
+
+procedure TKeymanControl.ShowKeyboardWelcome(const Keyboard: IKeymanKeyboardInstalled);
+var
+  pkg: IKeymanPackageInstalled;
+begin
+  pkg := Keyboard.OwnerPackage;
+  if Assigned(pkg) then
+    RunKeymanConfiguration('-kw "'+pkg.ID+'"');
+end;
+
+
+{procedure TKeymanVisualKeyboard.Print;  // I2329
+var
+  pkg: IKeymanPackageInstalled;
+  kbd: IKeymanKeyboardInstalled;
+begin
+  kbd := GetOwnerKeyboard;
+  if not Assigned(kbd) then Exit;
+  pkg := kbd.Get_OwnerPackage;
+  if Assigned(pkg) then
+    Context.Control.RunKeymanConfiguration('-kp "'+kbd.ID+'"');
+end;}
+
+
+{**
+  Control of startup and shutdown of keyman32.dll -- implemented only by 32 bit;
+  generally these functions are intended to be used only by keyman.exe.
+*}
+
+procedure TKeymanControl.StartKeyman32Engine;   // I5133
+begin
+{$IFDEF WIN64}
+  Error(Cardinal(E_NOTIMPL));
```

Comment:
I suppose this commit was establishing hooks for proper `WIN64` handling to be implemented in the later commits?

Reply:
Actually, in this case, not. The host for Keyman32 must be a 32-bit app so that it can run on 32-bit editions of Windows. Keyman32 is then responsible for starting the 64-bit version of the app on 64-bit editions of Windows and managing communication between the two engines.

Reply:
FWIW, it's likely that these APIs will only ever be used by keyman.exe. Originally the idea was multiple apps could host keyman32/keyman64. But the complexity of the host makes this less than attractive. We could definitely make a case for pulling these APIs out of `IKeymanControl` and into a new `IKeymanControlInternal` or something similar. Now is the time to make that change as we are still in Alpha... What do you think?

Reply:
That might be best.